src/os/bluestore/bluestore_types.cc

```diff
@@ -335,6 +335,241 @@ ostream& operator<<(ostream& out, const bluestore_extent_ref_map_t& m)
   return out;
 }
 
+// bluestore_blob_use_tracker_t
+
+void bluestore_blob_use_tracker_t::allocate()
+{
+  assert(num_au != 0);
+  bytes_per_au = new uint32_t[num_au];
+}
+
+void bluestore_blob_use_tracker_t::fall_back_to_partial(uint32_t _num_au, uint32_t _au_size)
+{
+  assert(num_au == 0 && au_size == 0);
+  assert(_num_au != 0 && _au_size != 0);
+  assert(_num_au * _au_size >= total_bytes);
+
+  uint32_t total = total_bytes;
+  num_au = _num_au;
+  au_size = _au_size;
+  allocate();
+  for (size_t i = 0; i < num_au; ++i) {
+    bytes_per_au[i] = min(total, au_size);
+    total -= bytes_per_au[i];
+  }
+  assert(total == 0);
+}
+
+void bluestore_blob_use_tracker_t::get(
+  uint32_t offset, uint32_t length,
+  std::function<void(uint32_t*, uint32_t*)> fallback_cb)
+{
+  if (!num_au) {
+    if (offset == 0 && total_bytes == 0) {
+      total_bytes = length;
+      return;
+    }
+    uint32_t _au_size = 0, _num_au = 0;
+    fallback_cb(&_num_au, &_au_size);
+    if (_num_au && _au_size) { // we might be forbidden to fallback to partial
```

Comment:
I'm afraid the use of total counting is more restrictive here: au_size <= blob size isn't a proper precondition to omit per_au counting. Consider the following case.
MAS = 16K.
write 8K~8K.
then change MAS to 4K
and write 0K~4K
One has to be able to switch to per-au counting at this point but that's impossible if per-au wasn't enabled at the first write.
Hence the only valid preconditions for total counting are compression or SINGLE 'get' (i.e. write) at OFFSET 0.